docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,728 @@
+        ret
+    .L.str:
+        .asciz  "%d"
+
+
+**GCC**
+
+.. code-block:: asm
+
+    .LC0:
+        .string "%d"
+    main:
+        push    rbx
+        xor     ebx, ebx
+    .L2:
+        mov     esi, ebx
+        mov     edi, OFFSET FLAT:.LC0
+        xor     eax, eax
+        add     ebx, 1
+        call    printf
+        cmp     ebx, 4
+        jne     .L2
+        xor     eax, eax
+        pop     rbx
+        ret
+
+
+**MSVC**
+
+.. code-block::
+
+    main    PROC
+        $LN12:
+        push    rbx
+        sub     rsp, 32
+        xor     ebx, ebx
+        npad    8
+    $LL4@main:
+        mov     edx, ebx
+        lea     rcx, OFFSET FLAT:'string'
+        call    printf
+        inc     ebx
+        cmp     ebx, 4
+        jl      SHORT $LL4@main
+        xor     eax, eax
+        add     rsp, 32
+        pop     rbx
+        ret     0
+    main    ENDP
+
+
+LLVM unrolls the the loop and compiles to a flat series of ``printf`` invocations while GCC and MSVC both agree to keep the loop intact, visible from the compare (``cmp``) and the jump (``jne``, ``jl``) instructions. LLVM code generation is identical to us having written the unrolled loop manually:
+
+.. code-block:: C++
+
+    printf("%d", 0);
+    printf("%d", 1);
+    printf("%d", 2);
+    printf("%d", 3);
+
+While there are various non-standard pragmas available to hint or force the compiler to unroll the loop, we instead use template meta-programming to force feed the compiler the unrolled loop.
+
+.. code-block:: C++
+
+    constexpr int size = 4;
+
+    // Maybe unrolled loop
+    for (int i = 0 ; i < size ; ++i)
+    {
+        printf("%d", i);
+    }
+
+    // Force unrolled loop
+    using namespace tmp;
+    static_for<0, less_than<size>, increment<1>>([]<int i>()
+    {
+        printf("%d", i);
+    });
+
+The most notable structural difference is that in the language ``for`` loop, the loop variable is given a name in the beginning, while in the ``static_for`` utility, the loop variable is given a name in the end. An important bonus is that in the body of the loop, you can use the running index ``i`` in contexts requiring constant expressions such as template arguments or inside ``if constexpr``.
+
+:code:`tmp::static_switch` takes runtime value and runtime dispatches to a range of set of tabulated functions, where said value is a compile-time constant and is eligible for use in compile-time evaluated contexts.
+
+Consider the following code:
+
+.. code-block:: C++
+
+    int warp_size = device_props.warpSize;
+    switch (warp_size)
+    {
+    case 32:
+        hipLaunchKernelGGL(kernel<32>, ...);
+        break;
+    case 64:
+        hipLaunchKernelGGL(kernel<64>, ...);
+        break;
+    }
+
+This all works fine as long as one doesn't commit copy-paste errors, as we had to repeat the possible values of ``warp_size`` our code is prepared to handle. This is what ``tmp::static_switch`` helps us with. The above is morally equivalent to:
+
```

Comment:
```suggestion
In the preceding code, note the code repetition for all possible values of ``warp_size``, the code is prepared to handle. To avoid this, use ``tmp::static_switch``, as shown:

```